internal/unionstore/art/art_snapshot.go

```diff
@@ -14,30 +14,111 @@
 
 package art
 
-import "context"
+import (
+	"context"
 
-func (*ART) SnapshotGetter() *SnapshotGetter {
-	panic("unimplemented")
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/internal/unionstore/arena"
+)
+
+func (t *ART) getSnapshot() arena.MemDBCheckpoint {
+	if len(t.stages) > 0 {
+		return t.stages[0]
+	}
+	return t.checkpoint()
+}
+
+// SnapshotGetter returns a Getter for a snapshot of MemBuffer.
+func (t *ART) SnapshotGetter() *SnapGetter {
+	return &SnapGetter{
+		tree: t,
+		cp:   t.getSnapshot(),
+	}
+}
+
+// SnapshotIter returns an Iterator for a snapshot of MemBuffer.
+func (t *ART) SnapshotIter(start, end []byte) *SnapIter {
+	inner, err := t.Iter(start, end)
+	if err != nil {
+		panic(err)
+	}
+	it := &SnapIter{
+		Iterator: inner,
+		cp:       t.getSnapshot(),
+	}
+	for !it.setValue() && it.Valid() {
+		_ = it.Next()
+	}
+	return it
 }
 
-func (*ART) SnapshotIter([]byte, []byte) *SnapshotIter {
-	panic("unimplemented")
+// SnapshotIterReverse returns a reverse Iterator for a snapshot of MemBuffer.
+func (t *ART) SnapshotIterReverse(k, lowerBound []byte) *SnapIter {
+	inner, err := t.IterReverse(k, lowerBound)
+	if err != nil {
+		panic(err)
+	}
+	it := &SnapIter{
+		Iterator: inner,
+		cp:       t.getSnapshot(),
+	}
+	for !it.setValue() && it.valid {
+		_ = it.Next()
+	}
+	return it
 }
 
-func (*ART) SnapshotIterReverse([]byte, []byte) *SnapshotIter {
-	panic("unimplemented")
+type SnapGetter struct {
+	tree *ART
+	cp   arena.MemDBCheckpoint
 }
 
-type SnapshotGetter struct{}
+func (snap *SnapGetter) Get(ctx context.Context, key []byte) ([]byte, error) {
+	addr, lf := snap.tree.search(key)
+	if addr.IsNull() {
+		return nil, tikverr.ErrNotExist
+	}
+	if lf.vAddr.IsNull() {
+		// A flags only key, act as value not exists
+		return nil, tikverr.ErrNotExist
+	}
+	v, ok := snap.tree.allocator.vlogAllocator.GetSnapshotValue(lf.vAddr, &snap.cp)
+	if !ok {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
 
-func (s *SnapshotGetter) Get(context.Context, []byte) ([]byte, error) {
-	panic("unimplemented")
+type SnapIter struct {
+	*Iterator
+	value []byte
+	cp    arena.MemDBCheckpoint
 }
 
-type SnapshotIter struct{}
+func (i *SnapIter) Value() []byte {
+	return i.value
+}
 
-func (i *SnapshotIter) Valid() bool   { panic("unimplemented") }
-func (i *SnapshotIter) Key() []byte   { panic("unimplemented") }
-func (i *SnapshotIter) Value() []byte { panic("unimplemented") }
-func (i *SnapshotIter) Next() error   { panic("unimplemented") }
-func (i *SnapshotIter) Close()        { panic("unimplemented") }
+func (i *SnapIter) Next() error {
+	i.value = nil
+	for i.Valid() {
+		if err := i.Iterator.Next(); err != nil {
+			return err
+		}
+		if i.setValue() {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (i *SnapIter) setValue() bool {
```

Comment:
For boolean returning value function, better to name it like `IsValueSet`.

Reply:
This function actually set the value, and the return boolean of this function is whether set value successful.

What about rename it to `trySetValue`?

Reply:
Perhaps `trySetValueRef` is better, also renaming the `value` field of `SnapIter` to `valueRef`, because it dose not own the value content.

Reply:
Not a big problem, `trySetValue` is ok too.